### PR TITLE
[Design] Example UseGoogleSignIn

### PR DESCRIPTION
**Pull request by HaoK**
Starting to play with the shape of the proposed Use/AddXyzSignIn...

```C#
services.AddAuthentication()
    // Pick one, last one will win (set to DefaultScheme) if more than one called (hence Use over Add)
   .UseGoogleSignIn(o => { })
   .UseFacebookSignIn(o => { })
   .UseOAuthSignIn(o => { })
   .UseOpenIdConnectSignIn(o => { })
   .UseTwitterSignIn(o => { })
   .UseIdentitySignIn()

// If you want to custom configure the cookie, you need to know which cookie
services.Configure<CookieAuthenticationOptions>(CookieAuthenticationDefaults.AuthenticationScheme, o => o.ExpireTimeSpan = 100);
// or with identity
services.ConfigureApplicationCookie(o => o.ExpireTimeSpan = 100);
```

Each Use is responsible for configuring the defaults (and adding any cookies as needed).  All of the social sign ins would use the default cookie.  Identity would probably continue to use its existing Identity.Application cookie as the default

We could also consider standardizing on a new constant "SignInCookie" as the scheme name for the cookie so these are all consistent, this would require some work for Identity to make it work, since SignInManager hardcodes IdentityConstants.ApplicationScheme today...

@Tratcher @davidfowl

### CODE REVIEW

File: samples/SocialSample/SignInExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+                // Override instead of default since this method is opinionated on the cookie scheme name.
+                o.SignInScheme = CookieAuthenticationDefaults.AuthenticationScheme;
+            });
+
+            builder.ConfigureCookie(GoogleDefaults.AuthenticationScheme);
+            return builder;
+        }
+
+        // Add the defalut cookie if needed, set it to default and configure the cookie to forward challenge
+        private static void ConfigureCookie(this AuthenticationBuilder builder, string forwardChallenge)
```
Comment (Tratcher) [inline samples/SocialSample/SignInExtensions.cs]: Aren't these also supposed to set cookies as the overall default scheme? Or how else does it work with `[Authroize]`

Comment (HaoK) [reply samples/SocialSample/SignInExtensions.cs]: It does on line 31:

`o.DefaultScheme = CookieAuthenticationDefaults.AuthenticationScheme;`

Comment (Tratcher) [reply samples/SocialSample/SignInExtensions.cs]: Ah, I mistook that for the new virtual DefaultScheme.

File: samples/SocialSample/SignInExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+                o.DefaultScheme = CookieAuthenticationDefaults.AuthenticationScheme;
+                if (!o.SchemeMap.ContainsKey(CookieAuthenticationDefaults.AuthenticationScheme))
+                {
+                    // TODO: make it easier to reuse AddCookie
+                    builder.Services.TryAddEnumerable(ServiceDescriptor.Singleton<IPostConfigureOptions<CookieAuthenticationOptions>, PostConfigureCookieAuthenticationOptions>());
+                    builder.AddScheme<CookieAuthenticationOptions, CookieAuthenticationHandler>(CookieAuthenticationDefaults.AuthenticationScheme, CookieAuthenticationDefaults.AuthenticationScheme, _ => { });
+                }
+            });
+            // Always configure the cookie's challenge
+            builder.Services.Configure<CookieAuthenticationOptions>(CookieAuthenticationDefaults.AuthenticationScheme, o => o.ForwardChallenge = forwardChallenge);
```
Comment (Tratcher) [inline samples/SocialSample/SignInExtensions.cs]: How is this better than setting DefaultChallengeScheme adjacent to DefaultScheme above?

Comment (HaoK) [reply samples/SocialSample/SignInExtensions.cs]: That only affects empty Challenge()...

This means all challenges that end up going to Cookies (including [Authorize]'s default), as well as Challenge("Cookies") go to google. which is what you want.

Comment (HaoK) [reply samples/SocialSample/SignInExtensions.cs]: For all intensive purposes this makes it impossible to challenge the cookie

**Comment by HaoK** [comment]
Replaced by https://github.com/aspnet/Security/pull/1644

**Pull request closed without merging**